chat/app.js

```diff
@@ -0,0 +1,176 @@
+const dgram = require('dgram');
+const readline = require('readline');
+const crypto = require('crypto');
+
+
+const Actions = {
+    message: 'message',
+    connect: 'connect',
+    check: 'check',
+    remove: 'remove'
+};
+
+const state = {
+    currentUser: '',
+    history: {},
+    companionPort: 0,
+    companionName: '',
+    currentHistory: null,
+    messenger: false,
+    client: null
+};
+
+const rl = readline.createInterface({
+    input: process.stdin,
+    output: process.stdout,
+    prompt: '>> '
+});
+
+rl.print = (line) => {
+    process.stdout.clearLine();
+    process.stdout.cursorTo(0);
+    console.log(line);
+    rl.prompt();
+}
+
+rl.prompt()
+
+rl.on('line', (line) => {
+    switch (line.trim()) {
+        case 'login':
+            login();
+            break;
+        case 'logout':
+            state.currentUser = '';
+            state.companionName = '';
+            state.companionPort = 0;
+            state.messenger = false;
+            state.client.close();
+            break;
+        case 'disconnect':
+            state.companionName = '';
+            state.companionPort = 0;
+            state.messenger = false;
+            break;
+        case 'connect':
+            connectToUser();
+            break;
+        case 'messenger':
+            if (state.companionPort) {
+                state.messenger = true;
+                let hash = generateHash([state.currentUser, state.companionName]);
+                state.history[hash].forEach(i => {
```

Comment:
So you are not recovering the history by sending it to the reconnected client? You just print messages stored in the memory instead?

Reply:
Yes, users save their conversations in their apps and, after reconnecting, restore them from their local history.
Do I need to send a history from the client to which the user is connected?

Reply:
I think so. Otherwise your history will still be lost when the app crashed.